File: data_structure/linked_list/src/LinkedListDemo.cs
// C#
// データ構造: 連結リスト (Linked List)

using System;
using System.Collections.Generic;

public class NodeData
{
    public object Data { get; set; }
    public NodeData Next { get; set; }

    public NodeData(object data)
    {
        Data = data;
        Next = null;
    }
}

public class LinkedListData
{
    private NodeData _data;
    private int _size;

    public LinkedListData()
    {
        _data = null;
        _size = 0;
    }

    public NodeData Get()
    {
        return _data;
    }

    public int GetPosition(object data)
    {
        if (IsEmpty())
            return -1;

        NodeData current = _data;
        int position = 0;

        while (current != null)
        {
            if (Equals(current.Data, data))
                return position;
            current = current.Next;
            position++;
        }

        return -1;
    }

    public object GetValue(int position)
    {
        if (IsEmpty() || position < 0 || position >= _size)
        {
            Console.WriteLine($"ERROR: {position} は範囲外です");
            return null;
        }

        NodeData current = _data;
        for (int i = 0; i < position; i++)
        {
            current = current.Next;
        }

        return current.Data;
    }

    public bool Add(object data, int? position = null)
    {
        NodeData newNode = new NodeData(data);

        if (IsEmpty())
        {
            _data = newNode;
            _size++;
            return true;
        }

        if (position == null || position >= _size)
        {
            NodeData current = _data;
            while (current.Next != null)
            {
                current = current.Next;
            }
            current.Next = newNode;
            _size++;
            return true;
        }

        if (position == 0)
        {
            newNode.Next = _data;
            _data = newNode;
            _size++;
            return true;
        }

        NodeData prev = _data;
        for (int i = 0; i < position - 1; i++)
        {
            prev = prev.Next;
        }

        newNode.Next = prev.Next;
        prev.Next = newNode;
        _size++;
        return true;
    }

    public bool Remove(int? position = null, object data = null)
    {
        if (IsEmpty())
        {
            Console.WriteLine("ERROR: リストが空です");
            return false;
        }

        if (data != null)
        {
            if (Equals(_data.Data, data))
            {
                _data = _data.Next;
                _size--;
                return true;
            }

            NodeData current = _data;
            while (current.Next != null && !Equals(current.Next.Data, data))
            {
                current = current.Next;
            }

            if (current.Next != null)
            {
                current.Next = current.Next.Next;
                _size--;
                return true;
            }
            else
            {
                Console.WriteLine($"ERROR: {data} は範囲外です");
                return false;
            }
        }

        if (position == null)
        {
            position = _size - 1;
        }

        if (position < 0 || position >= _size)
        {
            Console.WriteLine($"ERROR: {position} は範囲外です");
            return false;
        }

        if (position == 0)
        {
            _data = _data.Next;
            _size--;
            return true;
        }

        NodeData prev = _data;
        for (int i = 0; i < position - 1; i++)
        {
            prev = prev.Next;
        }

        prev.Next = prev.Next.Next;
        _size--;
        return true;
    }

    public bool Update(int position, object data)
    {
        if (IsEmpty() || position < 0 || position >= _size)
        {
            Console.WriteLine($"ERROR: {position} は範囲外です");
            return false;
        }

        NodeData current = _data;
        for (int i = 0; i < position; i++)
        {
            current = current.Next;
        }

        current.Data = data;
        return true;
    }

    public bool IsEmpty()
    {
        return _data == null;
    }

    public int Size()
    {
        return _size;
    }

    public bool Clear()
    {
        _data = null;
        _size = 0;
        return true;
    }

    public List<object> Display()
    {
        List<object> elements = new List<object>();
        NodeData current = _data;
        while (current != null)
        {
            elements.Add(current.Data);
            current = current.Next;
        }
        return elements;
    }
}

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("LinkedList TEST -----> start");

        Console.WriteLine("\nnew");
        LinkedListData linkedListData = new LinkedListData();
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nis_empty");
        bool output = linkedListData.IsEmpty();
        Console.WriteLine($"  出力値: {output}");

        Console.WriteLine("\nsize");
        int sizeOutput = linkedListData.Size();
        Console.WriteLine($"  出力値: {sizeOutput}");

        Console.WriteLine("\nadd");
        int input = 10;
        Console.WriteLine($"  入力値: {input}");
        bool addOutput = linkedListData.Add(input);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nadd");
        input = 20;
        Console.WriteLine($"  入力値: {input}");
        addOutput = linkedListData.Add(input);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nadd");
        object inputTuple = new { Data = 5, Position = 0 };
        Console.WriteLine($"  入力値: Data={((dynamic)inputTuple).Data}, Position={((dynamic)inputTuple).Position}");
        addOutput = linkedListData.Add(((dynamic)inputTuple).Data, ((dynamic)inputTuple).Position);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nadd");
        inputTuple = new { Data = 15, Position = 2 };
        Console.WriteLine($"  入力値: Data={((dynamic)inputTuple).Data}, Position={((dynamic)inputTuple).Position}");
        addOutput = linkedListData.Add(((dynamic)inputTuple).Data, ((dynamic)inputTuple).Position);
        Console.WriteLine($"  出力値: {addOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nget_value");
        int inputPosition = 1;
        Console.WriteLine($"  入力値: {inputPosition}");
        int positionOutput = linkedListData.GetPosition(inputPosition);
        Console.WriteLine($"  出力値: {positionOutput}");

        Console.WriteLine("\nget_value");
        input = 10;
        Console.WriteLine($"  入力値: {input}");
        positionOutput = linkedListData.GetPosition(input);
        Console.WriteLine($"  出力値: {positionOutput}");

        Console.WriteLine("\nupdate");
        dynamic updateInput = new { Position = 1, Data = 99 };
        Console.WriteLine($"  入力値: Position={updateInput.Position}, Data={updateInput.Data}");
        bool updateOutput = linkedListData.Update(updateInput.Position, updateInput.Data);
        Console.WriteLine($"  出力値: {updateOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nget_value");
        input = 15;
        Console.WriteLine($"  入力値: {input}");
        object valueOutput = linkedListData.GetValue(input);
        Console.WriteLine($"  出力値: {valueOutput}");

        Console.WriteLine("\nget_value");
        input = 100;
        Console.WriteLine($"  入力値: {input}");
        valueOutput = linkedListData.GetValue(input);
        Console.WriteLine($"  出力値: {valueOutput}");

        Console.WriteLine("\nremove");
        input = 15;
        Console.WriteLine($"  入力値: data={input}");
        bool removeOutput = linkedListData.Remove(data: input);
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nremove");
        int removePosition = 0;
        Console.WriteLine($"  入力値: position={removePosition}");
        removeOutput = linkedListData.Remove(position: removePosition);
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nremove");
        removeOutput = linkedListData.Remove();
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nremove");
        removePosition = 5;
        Console.WriteLine($"  入力値: position={removePosition}");
        removeOutput = linkedListData.Remove(position: removePosition);
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nclear");
        bool clearOutput = linkedListData.Clear();
        Console.WriteLine($"  出力値: {clearOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nis_empty");
        output = linkedListData.IsEmpty();
        Console.WriteLine($"  出力値: {output}");

        Console.WriteLine("\nsize");
        sizeOutput = linkedListData.Size();
        Console.WriteLine($"出力値: {sizeOutput}");

        Console.WriteLine("\nremove");
        removeOutput = linkedListData.Remove();
        Console.WriteLine($"  出力値: {removeOutput}");
        Console.WriteLine($"  現在のデータ: {string.Join(", ", linkedListData.Display())}");

        Console.WriteLine("\nLinkedList TEST <----- end");
    }
}

File: data_structure/linked_list/src/LinkedListDemo.java
// Java
// データ構造: 連結リスト (Linked List)

import java.util.ArrayList;
import java.util.List;

class NodeData {
    private Object data;
    private NodeData next;

    public NodeData(Object data) {
        this.data = data;
        this.next = null;
    }

    public Object getData() {
        return data;
    }

    public NodeData getNext() {
        return next;
    }
    
    public void setData(Object data) {
        this.data = data;
    }

    public void setNext(NodeData next) {
        this.next = next;
    }
}

class LinkedListData {
    private NodeData data;
    private int size;

    public LinkedListData() {
        this.data = null;
        this.size = 0;
    }

    public NodeData getData() {
        return this.data;
    }

    public int getPosition(Object searchData) {
        if (isEmpty()) {
            return -1;
        }
        
        NodeData current = this.data;
        int position = 0;
        
        while (current != null) {
            if (current.getData().equals(searchData)) {
                return position;
            }
            current = current.getNext();
            position++;
        }
        
        return -1;
    }

    public Object getValue(int position) {
        if (isEmpty() || position < 0 || position >= size) {
            System.out.printf("ERROR: %d は範囲外です%n", position);
            return null;
        }
        
        NodeData current = this.data;
        for (int i = 0; i < position; i++) {
            current = current.getNext();
        }
        
        return current.getData();
    }

    public boolean add(Object data) {
        return add(data, null);
    }

    public boolean add(Object data, Integer position) {
        NodeData newNode = new NodeData(data);
        
        if (isEmpty()) {
            this.data = newNode;
            size++;
            return true;
        }
        
        if (position == null || position >= size) {
            NodeData current = this.data;
            while (current.getNext() != null) {
                current = current.getNext();
            }
            current.setNext(newNode);
            size++;
            return true;
        }
        
        if (position == 0) {
            newNode.setNext(this.data);
            this.data = newNode;
            size++;
            return true;
        }
        
        NodeData current = this.data;
        for (int i = 0; i < position - 1; i++) {
            current = current.getNext();
        }
        
        newNode.setNext(current.getNext());
        current.setNext(newNode);
        size++;
        return true;
    }

    public boolean remove() {
        return remove(null, null);
    }

    public boolean remove(Integer position) {
        return remove(position, null);
    }

    public boolean remove(Object data) {
        return remove(null, data);
    }

    public boolean remove(Integer position, Object data) {
        if (isEmpty()) {
            System.out.println("ERROR: リストが空です");
            return false;
        }
        
        if (data != null) {
            if (this.data.getData().equals(data)) {
                this.data = this.data.getNext();
                size--;
                return true;
            }
            
            NodeData current = this.data;
            while (current.getNext() != null && !current.getNext().getData().equals(data)) {
                current = current.getNext();
            }
            
            if (current.getNext() != null) {
                current.setNext(current.getNext().getNext());
                size--;
                return true;
            } else {
                System.out.printf("ERROR: %s は範囲外です%n", data);
                return false;
            }
        }
        
        if (position == null) {
            position = size - 1;
        }
        
        if (position < 0 || position >= size) {
            System.out.printf("ERROR: %d は範囲外です%n", position);
            return false;
        }
        
        if (position == 0) {
            this.data = this.data.getNext();
            size--;
            return true;
        }
        
        NodeData current = this.data;
        for (int i = 0; i < position - 1; i++) {
            current = current.getNext();
        }
        
        current.setNext(current.getNext().getNext());
        size--;
        return true;
    }

    public boolean update(int position, Object data) {
        if (isEmpty() || position < 0 || position >= size) {
            System.out.printf("ERROR: %d は範囲外です%n", position);
            return false;
        }
        
        NodeData current = this.data;
        for (int i = 0; i < position; i++) {
            current = current.getNext();
        }
        
        current.setData(data);
        return true;
    }

    public boolean isEmpty() {
        return this.data == null;
    }

    public int size() {
        return this.size;
    }

    public boolean clear() {
        this.data = null;
        this.size = 0;
        return true;
    }
    
    public List<Object> display() {
        List<Object> elements = new ArrayList<>();
        NodeData current = this.data;
        while (current != null) {
            elements.add(current.getData());
            current = current.getNext();
        }
        return elements;
    }

    public static void main(String[] args) {
        System.out.println("LinkedList TEST -----> start");

        System.out.println("\nnew");
        LinkedListData linkedListData = new LinkedListData();
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nis_empty");
        boolean output = linkedListData.isEmpty();
        System.out.println("  出力値: " + output);

        System.out.println("\nsize");
        int sizeOutput = linkedListData.size();
        System.out.println("  出力値: " + sizeOutput);

        System.out.println("\nadd");
        int input = 10;
        System.out.println("  入力値: " + input);
        boolean addOutput = linkedListData.add(input);
        System.out.println("  出力値: " + addOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nadd");
        input = 20;
        System.out.println("  入力値: " + input);
        addOutput = linkedListData.add(input);
        System.out.println("  出力値: " + addOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nadd");
        int[] input1 = {5, 0};
        System.out.println("  入力値: " + input1[0] + ", " + input1[1]);
        addOutput = linkedListData.add(input1[0], input1[1]);
        System.out.println("  出力値: " + addOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nadd");
        int[] input2 = {15, 2};
        System.out.println("  入力値: " + input2[0] + ", " + input2[1]);
        addOutput = linkedListData.add(input2[0], input2[1]);
        System.out.println("  出力値: " + addOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nget_value");
        int findInput = 1;
        System.out.println("  入力値: " + findInput);
        int positionOutput = linkedListData.getPosition(findInput);
        System.out.println("  出力値: " + positionOutput);

        System.out.println("\nget_value");
        findInput = 10;
        System.out.println("  入力値: " + findInput);
        positionOutput = linkedListData.getPosition(findInput);
        System.out.println("  出力値: " + positionOutput);

        System.out.println("\nupdate");
        int[] updateInput = {1, 99};
        System.out.println("  入力値: " + updateInput[0] + ", " + updateInput[1]);
        boolean updateOutput = linkedListData.update(updateInput[0], updateInput[1]);
        System.out.println("  出力値: " + updateOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nget_value");
        int valueInput = 15;
        System.out.println("  入力値: " + valueInput);
        Object getValueOutput = linkedListData.getValue(valueInput);
        System.out.println("  出力値: " + getValueOutput);

        System.out.println("\nget_valuefind");
        valueInput = 100;
        System.out.println("  入力値: " + valueInput);
        getValueOutput = linkedListData.getValue(valueInput);
        System.out.println("  出力値: " + getValueOutput);

        System.out.println("\nremove");
        int removeInput = 15;
        System.out.println("  入力値: data=" + removeInput);
        boolean removeOutput = linkedListData.remove(removeInput);
        System.out.println("  出力値: " + removeOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nremove");
        int removePositionInput = 0;
        System.out.println("  入力値: position=" + removePositionInput);
        removeOutput = linkedListData.remove(removePositionInput);
        System.out.println("  出力値: " + removeOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nremove");
        removeOutput = linkedListData.remove();
        System.out.println("  出力値: " + removeOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nremove");
        removePositionInput = 5;
        System.out.println("  入力値: position=" + removePositionInput);
        removeOutput = linkedListData.remove(removePositionInput);
        System.out.println("  出力値: " + removeOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nclear");
        boolean clearOutput = linkedListData.clear();
        System.out.println("  出力値: " + clearOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nis_empty");
        output = linkedListData.isEmpty();
        System.out.println("  出力値: " + output);

        System.out.println("\nsize");
        sizeOutput = linkedListData.size();
        System.out.println("出力値: " + sizeOutput);

        System.out.println("\nremove");
        removeOutput = linkedListData.remove();
        System.out.println("  出力値: " + removeOutput);
        System.out.println("  現在のデータ: " + linkedListData.display());

        System.out.println("\nLinkedList TEST <----- end");
    }
}
